projects/plutus/src/current_sense.c

```diff
@@ -0,0 +1,111 @@
+#include "current_sense.h"
+
+#include <string.h>
+
+#include "critical_section.h"
+#include "wait.h"
+
+// Must be less than 128, or overflow may occur
+#define CURRENT_SENSE_RESET_SAMPLES 10
+
+static void prv_reset_samples(int32_t *value, void *context) {
+  CurrentSenseStorage *storage = (CurrentSenseStorage *)context;
+  if (storage->samples < CURRENT_SENSE_RESET_SAMPLES) {
+    storage->data->zero_point.voltage += *value;
+    storage->samples++;
+  }
+}
+
+static void prv_calculate_current(int32_t *value, void *context) {
+  CurrentSenseStorage *storage = (CurrentSenseStorage *)context;
+
+  // Formula for calculating calibrated current. Draws slope between given calibrated
+  // points, and uses the result as well as the voltage offset to calculate current
+  storage->current_value = storage->data->max_point.current *
+                           (*value - storage->data->zero_point.voltage) /
+                           (storage->data->max_point.voltage - storage->data->zero_point.voltage);
+  if (storage->callback != NULL) {
+    storage->callback(storage->current_value, storage->context);
+  }
+}
+
+StatusCode current_sense_init(CurrentSenseStorage *storage, CurrentSenseCalibrationData *data,
+                              LtcAdcStorage *adc_storage, LtcAdcSettings *settings) {
+  if (storage == NULL || adc_storage == NULL || settings == NULL) {
+    return status_code(STATUS_CODE_UNINITIALIZED);
+  }
+
+  // Initialize ADC and start periodic polling
+  storage->adc_storage = adc_storage;
+
+  status_ok_or_return(ltc_adc_init(storage->adc_storage, settings));
+  status_ok_or_return(
+      ltc_adc_register_callback(storage->adc_storage, prv_calculate_current, storage));
+
+  // Store calibration parameters
```

Comment:
You should be storing data and clearing storage before your register your callback, otherwise it may be called with invalid storage.

Reply:
Fair point, done